src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.cc

```diff
@@ -0,0 +1,604 @@
+
+#include <memory>
+#include <string.h>
+
+#include "include/buffer.h"
+#include "include/byteorder.h"
+#include "crimson/os/seastore/transaction_manager.h"
+#include "crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node.h"
+#include "crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.h"
+
+#include "test/crimson/seastore/test_block.h"   //for temporary
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::extentmap_manager {
+
+std::ostream &ExtMapInnerNode::print_detail(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+	     << ", depth=" << depth;
+}
+ExtMapInnerNode::seek_lextent_ret
+ExtMapInnerNode::seek_lextent(Transaction &t, objaddr_t lo, extent_len_t len)
+{
+  auto [begin, end] = bound(lo, lo + len);
+  auto result_up = std::make_unique<extent_map_list_t>();
+  auto &result = *result_up;
+  return crimson::do_for_each(
+    std::move(begin),
+    std::move(end),
+    [this, &t, &result, lo, len](const auto &val) mutable {
+      return extmap_load_extent(tm, t, val.get_val(), depth - 1).safe_then(
+	  [&t, &result, lo, len](auto extent) mutable {
+	    // TODO: add backrefs to ensure cache residence of parents
+	    return extent->seek_lextent(t, lo, len).safe_then(
+		[&t, &result, lo, len](auto item_list) mutable {
+		  result.splice(result.end(), item_list,
+				item_list.begin(), item_list.end());
+            });
+      });
+  }).safe_then([result=std::move(result_up)] {
+    return seek_lextent_ret(
+           seek_lextent_ertr::ready_future_marker{},
+           std::move(*result));
+  });
+}
+
+ExtMapInnerNode::insert_ret
+ExtMapInnerNode::insert(Transaction &t, objaddr_t lo, lext_map_val_t val)
+{
+  auto insertion_pt = get_containing_child(lo);
+  return extmap_load_extent(tm, t, insertion_pt->get_val(), depth-1).safe_then(
+    [this, insertion_pt, &t, lo, val=std::move(val)](auto extent) mutable {
+      return extent->at_max_capacity() ?
+        split_entry(t, lo, insertion_pt, extent) :
+        insert_ertr::make_ready_future<ExtMapNodeRef>(std::move(extent));
+    }).safe_then([this, &t, lo, val=std::move(val)](ExtMapNodeRef extent) mutable {
+      return extent->insert(t, lo, val);
+    });
+}
+
+ExtMapInnerNode::punch_lextent_ret
+ExtMapInnerNode::punch_lextent(Transaction &t, objaddr_t lo, extent_len_t len)
+{
+  auto result_up = std::make_unique<extent_map_list_t>();
+  auto &result = *result_up;
+  auto [begin, end] = bound(lo, lo + len);
+  logger().debug("ExtMapInnerNode::punch_lextent {} {}", lo, len);
+  return crimson::do_for_each(
+    std::move(begin),
+    std::move(end),
+    [this, &t, &result, lo, len](auto &punch_pt) mutable {
+    return extmap_load_extent(tm, t, punch_pt->get_val(), depth-1).safe_then(
+      [this, &result, punch_pt, &t, lo, len](auto extent) mutable {
+      return extent->at_max_capacity() ?
+        split_entry(t, lo, punch_pt, extent) :
```

Comment:
It would appear that punching a hole could result in an extra entry, but it seems like it would be even more common that it results in fewer.  Shouldn't you also need to check for merge?